<a href="https://colab.research.google.com/github/sanket143/Notebooks/blob/master/Hacker_News_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Hacker News Analysis

This notebook compares the number of comments on "Ask HN" and "Show HN" posts.

In [0]:
from csv import reader

hacker_news_file = open("hacker_news.csv");
hn = list(reader(hacker_news_file));

print(hn[:5])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [0]:
headers = hn[0]
hn = hn[1:]

print(headers)
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']
[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [0]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(row);
    elif title.lower().startswith("show hn"):
        show_posts.append(row);
    else:
        other_posts.append(row);

print(len(ask_posts), len(show_posts), len(other_posts));

1744 1162 17194


In [0]:
print(ask_posts[:5])
print(show_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]
[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h

In [0]:
total_ask_comments = 0
total_show_comments = 0

for row in ask_posts:
    total_ask_comments += int(row[4])

avg_ask_comments = total_ask_comments / len(ask_posts)

for row in show_posts:
    total_show_comments += int(row[4])
    
avg_show_comments = total_show_comments / len(show_posts)

print(avg_ask_comments, avg_show_comments)

14.038417431192661 10.31669535283993


Avg Ask  HN posts: 14.038
Avg Show HN posts: 10.31

People comment on "Ask HN" posts more than "Show HN"

In [0]:
import datetime as dt

result_list = []

for row in ask_posts:
    result_list.append([row[6], int(row[4])])
    
counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    post_date = dt.datetime.strptime(row[0], "%m/%d/%Y %H:%M")
    hour = post_date.strftime("%H");
    
    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1;
        comments_by_hour[hour] = row[1];
    else:
        counts_by_hour[hour] += 1;
        comments_by_hour[hour] += row[1]


In [0]:
avg_by_hour = [];

for hour in counts_by_hour:
    avg = comments_by_hour[hour] / counts_by_hour[hour];
    avg_by_hour.append([hour, avg])

print(avg_by_hour)

[['05', 10.08695652173913], ['17', 11.46], ['15', 38.5948275862069], ['06', 9.022727272727273], ['11', 11.051724137931034], ['00', 8.127272727272727], ['23', 7.985294117647059], ['18', 13.20183486238532], ['07', 7.852941176470588], ['16', 16.796296296296298], ['08', 10.25], ['01', 11.383333333333333], ['14', 13.233644859813085], ['12', 9.41095890410959], ['13', 14.741176470588234], ['21', 16.009174311926607], ['22', 6.746478873239437], ['20', 21.525], ['10', 13.440677966101696], ['04', 7.170212765957447], ['19', 10.8], ['09', 5.5777777777777775], ['02', 23.810344827586206], ['03', 7.796296296296297]]


In [0]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])
    
print(swap_avg_by_hour)

sorted_swap = sorted(swap_avg_by_hour, reverse=True)


[[10.08695652173913, '05'], [11.46, '17'], [38.5948275862069, '15'], [9.022727272727273, '06'], [11.051724137931034, '11'], [8.127272727272727, '00'], [7.985294117647059, '23'], [13.20183486238532, '18'], [7.852941176470588, '07'], [16.796296296296298, '16'], [10.25, '08'], [11.383333333333333, '01'], [13.233644859813085, '14'], [9.41095890410959, '12'], [14.741176470588234, '13'], [16.009174311926607, '21'], [6.746478873239437, '22'], [21.525, '20'], [13.440677966101696, '10'], [7.170212765957447, '04'], [10.8, '19'], [5.5777777777777775, '09'], [23.810344827586206, '02'], [7.796296296296297, '03']]


In [0]:
print("Top 5 Hours for Ask Posts Comments")

for row in sorted_swap[:5]:
    template = "{}:00: {:.2f} average comments per post"
    summary = template.format(row[1], float(row[0]))
    print(summary)
    

Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post
